src/nix/path-info.cc

```diff
@@ -87,41 +96,98 @@ struct CmdPathInfo : StorePathsCommand, MixJSON
             pathLen = std::max(pathLen, store->printStorePath(storePath).size());
 
         if (json) {
-            std::cout << store->pathInfoToJSON(
+            auto json = store->pathInfoToJSON(
                 // FIXME: preserve order?
                 StorePathSet(storePaths.begin(), storePaths.end()),
-                true, showClosureSize, SRI, AllowInvalid).dump();
+                true, showClosureSize, SRI, AllowInvalid);
+
+            if (showSubStatus) {
+                ThreadPool pool(fileTransferSettings.httpConnections);
+
+                Sync<nlohmann::json> jsonList_;
+
+                {
+                    auto jsonList(jsonList_.lock());
+                    *jsonList = nlohmann::json::array();
+                }
+
+                std::function<void(nlohmann::json::basic_json::value_type)> checkSubs;
+                checkSubs = [&](const nlohmann::json::basic_json::value_type & value) {
+                    auto clone = value;
+                    std::string storePathS = value["path"];
+                    auto info = store->queryPathInfo(store->followLinksToStorePath(storePathS));
+                    auto storePath = info->path;
+
+                    SubstitutablePathInfos infos;
+                    store->querySubstitutablePathInfos({{storePath, std::nullopt}}, infos);
+
+                    clone["substitute"] = !infos.empty();
+
+                    auto jsonList(jsonList_.lock());
+                    jsonList->emplace_back(clone);
+                };
+
+                for (auto v : json) {
+                    pool.enqueue(std::bind(checkSubs, v));
+                }
+                pool.process();
+
+                auto jsonList(jsonList_.lock());
+                std::cout << jsonList->dump();
+            }
+            else std::cout << json.dump();
         }
 
         else {
+            ThreadPool pool(fileTransferSettings.httpConnections);
 
-            for (auto & storePath : storePaths) {
+            Sync<std::set<std::string>> results_;
+
+            std::function<void(StorePath)> printPath;
+            printPath = [&](const StorePath & storePath) {
```

Comment:
Here, the parameter type from the `std::function` and the actual assignment differ by having and not having `const ... &` qualifiers. i suggest using the same nested lambda expression approach here, too.